CRB-182 Test for containers read-only filesystem

atodorov opened the pull request:
we don't want containers to modify any files inside of the image once they are up and running.

This should also be supplemented with individual checks for each container. They could be added after the proposed health check changes.

atodorov committed:
Test for containers read-only filesystem

we don't want containers to modify any files inside of the image once
they are up and running.

CAGS295 reviewed (approved):
Correct me if I'm wrong, we still need to make containers read only. This PR tests for image diffs, does this affect JIT compilation and runtimes that cache metadata into disk? For example, python may generate .pyc's while executing regular .py files.

colingluwa commented:
Please review the following blog post. 
https://benlobaugh.medium.com/how-to-secure-docker-containers-with-a-read-only-filesystem-b27230f6efb

The docker service inside the docker file should contain the following flag. 

read_only: true


You may need to identify certain directories that require write permissions.

atodorov committed:
Mark services as read_only: true

and mark certain directories as writeable.

will-gluwa commented on Client/docker-compose.yaml:
```diff
@@ -22,6 +22,7 @@ services:
     network_mode: "host"
     image: gluwa/creditcoin-client:1.7
     container_name: creditcoin-client
+    read_only: true
```
Have you tested this works? My assumption is that since the client tries to write to plugins/progress.txt when a command is run this will cause unexpected behaviour.

cc: @vladimir-gluwa

colingluwa replied:
Correct, if the client tries to write to a location that is not identified as a writable location it should cause the application to fail. 

This should be added to the documentation, all the file system locations that we are reading/writing.

atodorov replied:
I've listed the plugins directory as tmpfs in the config file and added several more checks. However if we really need fully working containers that means we need to configure credentials inside GitHub secrets and at some point these are bound to leak. 

How do we want to handle that and keeping the repository public?

colingluwa replied:
Any deployment that requires credentials or secrets will have those secrets injected at deployment time. 
This is just so that the systems within the container cannot be modified or manipulated.

atodorov committed:
Add more container sanity checks

When in read_only mode containers should not:

- exit
- basic commands (e.g. --help) should work
